common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java

```diff
@@ -331,9 +357,33 @@ public MicrosoftStsTokenRequest createTokenRequest(@NonNull final MicrosoftStsAu
         tokenRequest.setRedirectUri(request.getRedirectUri());
         tokenRequest.setClientId(request.getClientId());
         tokenRequest.setScope(request.getTokenScope());
-
         tokenRequest.setGrantType(TokenRequest.GrantTypes.AUTHORIZATION_CODE);
+        setTokenRequestCorrelationId(tokenRequest);
+
+        try {
+            if (SCHEME_POP.equals(mStrategyOptions.getAuthenticationScheme().getName())) {
+                // Add a token_type
+                tokenRequest.setTokenType(TokenRequest.TokenType.POP);
+
+                // Generate keys if they don't already exist...
+                if (!mDevicePopManager.asymmetricKeyExists()) {
+                    mDevicePopManager.generateAsymmetricKey();
+                }
+
+                // Set the req_cnf
+                final String reqCnf = mDevicePopManager.getRequestConfirmation();
+
+                tokenRequest.setRequestConfirmation(reqCnf);
+            }
+        } catch (final ClientException e) {
+            // TODO how should this be handled?
+            e.printStackTrace();
```

Comment:
TODO Make sure these exceptions are surfaced such that they can be caught/passed back to the caller

Reply:
Resolved in c4b7751927824d511a000741c83d5f5014fdf016